src/dotnet/commands/dotnet-msbuild/MSBuildForwardingApp.cs

```diff
@@ -36,7 +46,16 @@ public MSBuildForwardingApp(IEnumerable<string> argsToForward)
 
         public int Execute()
         {
-            return _forwardingApp.Execute();
+            try
+            {
+                Environment.SetEnvironmentVariable(TelemetrySessionIdEnvironmentVariableName, Telemetry.CurrentSessionId);
```

Comment:
As an example of this, CLI build invokes <Exec> of CLI all over the place. This may be a common pattern, and the env var automatically gets plumbed through to child processes. Are the issues you're referring to largely in unit tests clashing with one another?